crates/ibc/src/core/ics02_client/client_state.rs

```diff
@@ -51,15 +52,71 @@ pub trait ClientState:
     /// state timestamp
     fn expired(&self, elapsed: Duration) -> bool;
 
-    /// Convert into a boxed trait object
-    fn into_box(self) -> Box<dyn ClientState>
-    where
-        Self: Sized,
-    {
-        Box::new(self)
-    }
+    /// Verify the upgraded client and consensus states and validate proofs
+    /// against the given root.
+    ///
+    /// NOTE: proof heights are not included as upgrade to a new revision is
+    /// expected to pass only on the last height committed by the current
+    /// revision. Clients are responsible for ensuring that the planned last
+    /// height of the current revision is somehow encoded in the proof
+    /// verification process. This is to ensure that no premature upgrades
+    /// occur, since upgrade plans committed to by the counterparty may be
+    /// cancelled or modified before the last planned height.
+    fn verify_upgrade_client(
+        &self,
+        upgraded_client_state: Any,
+        upgraded_consensus_state: Any,
+        proof_upgrade_client: MerkleProof,
+        proof_upgrade_consensus_state: MerkleProof,
+        root: &CommitmentRoot,
+    ) -> Result<(), ClientError>;
 
-    fn initialise(&self, consensus_state: Any) -> Result<Box<dyn ConsensusState>, ClientError>;
+    // Verify_membership is a generic proof verification method which verifies a
+    // proof of the existence of a value at a given Path.
+    fn verify_membership(
+        &self,
+        prefix: &CommitmentPrefix,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        path: Path,
+        value: Vec<u8>,
+    ) -> Result<(), ClientError>;
+
+    // Verify_non_membership is a generic proof verification method which
+    // verifies the absence of a given commitment.
+    fn verify_non_membership(
+        &self,
+        prefix: &CommitmentPrefix,
+        proof: &CommitmentProofBytes,
+        root: &CommitmentRoot,
+        path: Path,
+    ) -> Result<(), ClientError>;
+}
+
+/// `ClientState` methods which require access to the client's
+/// `ValidationContext`.
+///
+/// The `ClientValidationContext` enables the light client implementation to
+/// define its own `ValidationContext` trait and use it in its implementation.
+///
+/// ```ignore
+/// impl<ClientValidationContext> ClientStateValidation<ClientValidationContext> for MyClientState
+/// where
+///     ClientValidationContext: MyValidationContext,
+/// {
+///   // `MyValidationContext` methods available
+/// }
+///
+/// trait MyValidationContext {
+///   // My Context methods
+/// }
+/// ```
+pub trait ClientStateValidation<ClientValidationContext> {
```

Comment:
> Not sure if we could effectively restrain the light client capabilities here (I mean the [3rd goal](https://github.com/cosmos/ibc-rs/pull/708/files#diff-f0fc94f4a7703342248727cb915b38b67295163eaf33e0696401cde405c8ec23R12-R14), stated in ADR07).
Now it is the responsibility of light client developers to ensure proper restriction of generics,

Our threat model is not "light client developers can potentially be malicious", as we clearly state that host chain developers should audit all the light client code they use. Then, in our design, we can assume non-malicious light clients (as malicious ones will not be included in any real chain). Hence, we don't need to try to restrict in any way the shape of traits that light clients use. Note that if a light client is going to be malicious, it can very well do it in a "normal looking API". The only small advantage that the current design has over the previous one in that respect is that we don't force light client developers to use a fixed set of methods, which they might misunderstand and misuse (hence leading to incorrect code).

Reply:
No, I didn’t mean restraining b/c of malicious behaviour. was thinking the same way… in terms of better development experience, reducing method misusage and being more specific of what a context argument can offer.